boskos/boskos.go

```diff
@@ -0,0 +1,295 @@
+	if err != nil {
+		logrus.WithError(err).Fatal("Fail to create my ranch!")
+	}
+
+	http.Handle("/", handleDefault())
+	http.Handle("/request", handleRequestProject(r))
+	http.Handle("/return", handleReturnProject(r))
+	http.Handle("/sync", handleHeartbeat(r))
+
+	logrus.Info("Start Service")
+	logrus.WithError(http.ListenAndServe(":8080", nil)).Fatal("ListenAndServe returned.")
+}
+
+func handleDefault() http.HandlerFunc {
+	return func(res http.ResponseWriter, req *http.Request) {
+		logrus.WithField("handler", "handleDefault").Infof("From %v", req.RemoteAddr)
+	}
+}
+
+func handleRequestProject(r *RanchoSantaFe) http.HandlerFunc {
+	return func(res http.ResponseWriter, req *http.Request) {
+		logrus.WithField("handler", "handleRequestProject").Infof("From %v", req.RemoteAddr)
+
+		// validate vars
+		job := req.URL.Query().Get("job")
+		durationStr := req.URL.Query().Get("duration")
+		if job == "" || durationStr == "" {
+			logrus.Warning("[BadRequest]Job and duration must present in the request")
+			http.Error(res, "Job and duration must present in the request", http.StatusBadRequest)
+			return
+		}
+
+		duration, err := time.ParseDuration(durationStr)
+		if err != nil {
+			http.Error(res, "Invalid Duration", http.StatusBadRequest)
+			logrus.WithError(err).Errorf("Bad duration: %v", durationStr)
+			return
+		}
+
+		logrus.Infof("Job %v request for a project for %v", job, durationStr)
+
+		proj := r.GimmeCattle(job, duration)
+
+		if proj != nil {
+			projJSON, err := json.Marshal(proj)
+			if err != nil {
+				logrus.WithError(err).Errorf("json.Marshal failed : %v", proj)
+				http.Error(res, error.Error(err), http.StatusInternalServerError)
+				return
+			}
+			logrus.Infof("Job %v checked out %v", job, string(projJSON))
+			fmt.Fprint(res, string(projJSON))
+		} else {
+			logrus.Infof("No available project")
+			res.WriteHeader(http.StatusNoContent)
+			fmt.Fprint(res, "No available project, please try again later\n")
+		}
+	}
+}
+
+func handleReturnProject(r *RanchoSantaFe) http.HandlerFunc {
+	return func(res http.ResponseWriter, req *http.Request) {
+		logrus.WithField("handler", "handleReturnProject").Infof("From %v", req.RemoteAddr)
+
+		// validate vars
+		job := req.URL.Query().Get("job")
+		if job == "" {
+			http.Error(res, "job must present in the request", http.StatusBadRequest)
+			return
+		}
+
+		logrus.Infof("Job %v returns a project", job)
+		if err := r.ReturnCattle(job); err == nil {
+			fmt.Fprint(res, "Successful\n")
+		} else {
+			logrus.WithError(err).Error("Return failed")
+			http.Error(res, error.Error(err), http.StatusInternalServerError)
+		}
+	}
+}
+
+func handleHeartbeat(r *RanchoSantaFe) http.HandlerFunc {
+	return func(res http.ResponseWriter, req *http.Request) {
+		logrus.WithField("handler", "handleHeartbeat").Infof("From %v", req.RemoteAddr)
+		if !r.CheckLease() {
+			res.WriteHeader(http.StatusNoContent)
+		}
+	}
+}
+
+type RanchoSantaFe struct {
+	MyCattle    []*GCPProject
```

Comment:
I enjoy creative names and believe that it is important for developers to be able to express themselves.

However, I am hesitant to embed the cattle metaphor so deeply into this project. Especially in light of the fact that we spent a bunch of effort renaming pet set.

Is there an alternative way we can represent things here?